BRP registry JSON schema endpoint

Leinnan opened the pull request:
# Objective

Resolve #16745

## Solution

Provide a way to map `AppTypeRegistry` types into a JSON Schema that can be used in other applications. I took code from https://github.com/kaosat-dev/Blenvy as a starting point, cleaned up and adapter more for `bevy_remote` needs. Based on feedback and needs it could be improved, I could add some filtering options, etc.

## Testing

- I was comparing results with the ones from code in `blenvy`
- There is added unit test, could be added more
- I was testing it in my game with this code:
```rust
fn types_to_file(world: &mut World) {
    use bevy_remote::builtin_methods::export_registry_types;
    let Ok(Ok(types_schema)) = world.run_system_cached_with(export_registry_types, None) else {
        return;
    };
    let registry_save_path = std::path::Path::new("assets").join("registry.json");
    let writer =
        std::fs::File::create(registry_save_path).expect("should have created schema file");
    serde_json::to_writer_pretty(writer, &types_schema).expect("Failed to save types to file");
}
```
It can be run by adding it at startup 
```rust
app.add_systems(Startup, types_to_file);
```

NthTensor commented:
Do you think it will be possible to do the inverse as well eventually? Take a serialized map and load it into the registry?

MrGVSV commented on crates/bevy_remote/src/builtin_methods.rs:
```diff
@@ -1017,4 +1242,37 @@ mod tests {
     }
+
+    #[test]
+    fn reflect_export_test() {
```
It'd be great if we had some more tests for different kinds of types (e.g. enums, tuples, etc.).

Would also be helpful if one of the tests also tested the JSON output (i.e. compared the entire output to a string). This also helps to ensure that everything not only behaves correctly, but _looks_ right too (and it's nice for for future devs to see what the result is actually supposed to look like).

Leinnan replied:
Agree, I will try to add some more tests.

Leinnan commented:
I mean, maybe? I don't have too big knowledge about reflect in bevy but AFAIK there is a option to register type by providing `TypeRegistration` into the `TypeRegistry` (`add_registration` method). `TypeRegistration` is made up from `TypeInfo`, the same from which the JSON Schemas are constructed and the `dyn TypeData`. ATM you can create `TypeRegistration` by passing type like `TypeRegistration::of::<Option<String>>();`, but maybe in the future there could be a way to transform JSON Schema into `TypeRegistration` object that could be registered.

MrGVSV commented:
> Do you think it will be possible to do the inverse as well eventually? Take a serialized map and load it into the registry? Maybe @MrGVSV knows.

I actually like the idea of having this sort of endpoint be separate from `bevy_reflect` for now. It allows BRP work to be unblocked by `bevy_reflect` work and also allows it to be more flexible in what data it collects/cares about.

Creating the registry the other way around is going to be a bit more complex, involves working it into `bevy_reflect`, and probably requires a rethink of how we register/recognize types and their info. Some examples of changes that may be needed:

- Completely moving away from `TypeId` as the de facto type key (favoring `TypePath` instead)
- Change `TypeInfo` to no longer be `&'static` (or work out a way to properly intern it)
- Create a split between compile-time `TypeInfo` and runtime `TypeInfo` (probably as two separate types), specifically so we can store function pointers in `TypeInfo`

So yeah, because of the complexity involved, I think it's better to work out solutions in `bevy_remote` instead of waiting on `bevy_reflect` to catch up.

NthTensor commented:
The reason I ask is because bsn is reflection-backed and (as far as I can tell) can't even be parsed when it uses a type that's not defined locally. Alice was wondering if this would let us get around that. From that point of view, doing stuff entirely on BRP won't be useful unless bsn is backed by BRP instead of reflection. And that isn't on the table.

Leinnan commented on crates/bevy_remote/src/builtin_methods.rs:
```diff
@@ -801,6 +804,228 @@ pub fn process_remote_list_watching_request(
+                "long_name": t.type_path(),
+                "properties": properties,
+                "additionalProperties": false,
+                "required": info
+                    .iter()
+                    .filter(|field| !field.type_path().starts_with("core::option::Option"))
+                    .map(NamedField::name)
+                    .collect::<Vec<_>>(),
```
It is part of the standard: https://json-schema.org/draft/2020-12/json-schema-validation#section-6.5.3

Leinnan commented:
@MrGVSV For naming I think we should just follow https://json-schema.org/understanding-json-schema/keywords when possible. 

But we are providing here some values that are not definied in JSON Schema, this is mostly the ones with snake case, so I think for those we could try to follow Bevy naming scheme, but make them camel case so they are consistent with the rest of the fields.

Leinnan commented on crates/bevy_remote/src/builtin_methods.rs:
```diff
@@ -801,6 +804,228 @@ pub fn process_remote_list_watching_request(
+                json!({
+                    "type": "string",
+                    "typeInfo": "Enum",
+                    "long_name": t.type_path(),
+                    "oneOf": info
```
Responded here: https://github.com/bevyengine/bevy/pull/16882#issuecomment-2552221182

Leinnan commented on crates/bevy_remote/src/builtin_methods.rs:
```diff
@@ -801,6 +804,228 @@ pub fn process_remote_list_watching_request(
+                .collect::<Map<_, _>>();
+            json!({
+                "type": "object",
+                "typeInfo": "Struct",
+                "long_name": t.type_path(),
+                "properties": properties,
```
Responded here: https://github.com/bevyengine/bevy/pull/16882#issuecomment-2552221182

MrGVSV commented:
> The reason I ask is because bsn is reflection-backed and (as far as I can tell) can't even be parsed when it uses a type that's not defined locally. Alice was wondering if this would let us get around that. From that point of view, doing stuff entirely on BRP won't be useful unless bsn is backed by BRP instead of reflection. And that isn't on the table.

I see. We can definitely work towards it, but I wasn't planning on trying to get these sort of major changes into 0.16 since reflection just had a massive overhaul in 0.15, and this would likely be another major change. My current goal for 0.16 is just cleanup work, general improvements/fixes, and maybe some cool mostly-additive features (maybe some relatively small breaking changes).

We could try to do it, but ideally most of the work would be completed and merged early into the cycle, and we're already a month in with no progress made yet 😅

NthTensor commented:
Yeah no worries, just wondering about how much effort it would take. After some further thought, it's probably simpler just to write a second bsn parser that does use brp/schemas instead of reflection, as you suggested.

Leinnan commented on crates/bevy_remote/src/builtin_methods.rs:
```diff
@@ -801,6 +804,228 @@ pub fn process_remote_list_watching_request(
+                    VariantInfo::Struct(v) => json!({
+                        "type": "object",
+                        "typeInfo": "Struct",
+                        "long_name": v.name(),
+                        "short_name": v.name().split("::").last().unwrap_or(v.name()),
```
Yeah, it doesn't seems to make sense, could be a result of moving code from some of the previous versions?

Leinnan replied:
Updated, now it's the other way around, I am adding base type path to the variant and combine them:
```rust
                        "typePath": format!("{}::{}", type_path, v.name()),
                        "shortPath": v.name(),
```

Leinnan commented:
I did update branch quite a bit, most important changes:
- created `JsonSchemaBevyType` struct, so the structure is more defined and later on it could be easier to add documentation and make the type public.
- naming pass: 
    - `short_name` -> `shortPath`
    - `typeInfo` -> `kind`
    - `long_name` -> `typePath`
- two boolean fields are only serialized when value is true: `isComponent` and `isResource`.

Leinnan commented:
I've updated structs documentation, added filtering options to the query, removed `isComponent` and `isResource` fields and added new `reflectTypes` vec field that can contain info about reflect types that type is providing. 

Example:
Rust definition:
```rust
        #[derive(
            bevy_reflect::Reflect, bevy_ecs::system::Resource, Default, Deserialize, Serialize,
        )]
        #[reflect(Resource, Default, Serialize, Deserialize)]
        struct Foo {
            a: f32,
        }
```

Type schema:
```json
{
  "shortPath": "Foo",
  "typePath": "bevy_remote::builtin_methods::tests::Foo",
  "modulePath": "bevy_remote::builtin_methods::tests",
  "crateName": "bevy_remote",
  "reflectTypes": [
    "Resource",
    "Default",
    "Serialize",
    "Deserialize"
  ],
  "kind": "Struct",
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "a": {
      "type": {
        "$ref": "#/$defs/f32"
      }
    }
  },
  "required": [
    "a"
  ]
}
```

Leinnan commented:
@MrGVSV IMHO now besides the more tests it is quite done, let me know how you feel about it.

NthTensor reviewed (approved):
This looks great!
